packages/blocks/src/api/registration.js

```diff
@@ -758,3 +758,178 @@ export const registerBlockVariation = ( blockName, variation ) => {
 export const unregisterBlockVariation = ( blockName, variationName ) => {
 	dispatch( blocksStore ).removeBlockVariations( blockName, variationName );
 };
+
+/**
+ * Registers a new block bindings source with an object defining its
+ * behavior. Once registered, the source is available to be connected
+ * to the supported block attributes.
+ *
+ * @param {Object}   source                    Properties of the source to be registered.
+ * @param {string}   source.name               The unique and machine-readable name.
+ * @param {string}   source.label              Human-readable label.
+ * @param {Function} [source.getValue]         Function to get the value of the source.
+ * @param {Function} [source.setValue]         Function to update the value of the source.
+ * @param {Function} [source.setValues]        Function to update multiple values connected to the source.
+ * @param {Function} [source.getPlaceholder]   Function to get the placeholder when the value is undefined.
+ * @param {Function} [source.canUserEditValue] Function to determine if the user can edit the value.
+ *
+ * @example
+ * ```js
+ * import { _x } from '@wordpress/i18n';
+ * import { registerBlockBindingsSource } from '@wordpress/blocks'
+ *
+ * registerBlockBindingsSource( {
+ *     name: 'plugin/my-custom-source',
+ *     label: _x( 'My Custom Source', 'block bindings source' ),
+ *     getValue: () => 'Value to place in the block attribute',
+ *     setValue: () => updateMyCustomValue(),
+ *     setValues: () => updateMyCustomValuesInBatch(),
+ *     getPlaceholder: () => 'Placeholder text when the value is undefined',
+ *     canUserEditValue: () => true,
+ * } );
+ * ```
+ */
+export const registerBlockBindingsSource = ( source ) => {
+	const {
+		name,
+		label,
+		getValue,
+		setValue,
+		setValues,
+		getPlaceholder,
+		canUserEditValue,
+	} = source;
+
+	// Check if the source is already registered.
+	const existingSource = unlock(
+		select( blocksStore )
+	).getBlockBindingsSource( name );
+	if ( existingSource ) {
+		console.error(
```

Comment:
Should this be translatable and dev only? Can be done in a follow-up.

Reply:
Maybe? 🤔 I checked the rest of the `console.error` and `console.warn` Gutenberg has and it seems they are never translated. Maybe it is a good discussion to have globally? By the way, what do you mean with dev only?

Reply:
The most optimal would be:

```js
import { warning } from `@wordpress/warning`; 
warning( 'This is the message...' );
```

It's a hint for a developer, but the code continues to work after making the best informed decision.

Reply:
I can make that change 🙂 

I just followed what is being done in the `registerBlockType` function: [link](https://github.com/WordPress/gutenberg/blob/e6e7bb05206a94e56cd9f27405a54ae99faf7758/packages/blocks/src/api/registration.js#L227-L242).

Reply:
I am not sure when we should throw a warning and when an error. In both cases, the editor won't break but the bindings won't work as expected.

**Error**
![Screenshot 2024-07-16 at 11 59 55](https://github.com/user-attachments/assets/8cd754ee-e8db-4040-838e-fa7ac9644bc7)


**Warning**
![Screenshot 2024-07-16 at 11 59 10](https://github.com/user-attachments/assets/fc94caaa-86bf-4a25-a507-d4f750ed2b60)

I'll create a follow-up pull request to properly discuss that.

Reply:
That's a good question. Interactivity API uses warnings, whether the interactivity worked or not.

Reply:
It would be great to unify. `@wordpress/warning` was introduced not that long ago, so historically, people would use `console.error` or `console.warn` based on their personal preferences. I would simplify it and use `warning` from `@wordpress/warning` as it doesn't add too much value on production anyway. The other benefit is that `@wordpress/warning` uses hooks internally so 3rd party plugin can intercept these warning. If I recall correctly @johnbillion integrated that into [Query Monitor plugin](https://wordpress.org/plugins/query-monitor/) that has 200k active installs.

Reply:
That makes sense. I will create a pull request to change all the `console.error` and `console.warn` in the registration file to use `@wordpress/warning`.

Reply:
Follow-up created here: https://github.com/WordPress/gutenberg/pull/63610. Although I encountered some issues.